sklearn/utils/validation.py

```diff
@@ -1018,6 +1019,151 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+      1e-5 times the maximum real part). If this check fails, it raises a
+      ``ValueError``. Otherwise all non-significant imaginary parts that may
+      remain are removed.
+
+    - that eigenvalues are not all negative. If this check fails, it raises a
+      ``ValueError``
+
+    - that there are no significant negative eigenvalues (with absolute value
+      more than 1e-10 and more than 1e-5 times the largest positive
+      eigenvalue). If this check fails, it raises a
+      ``PositiveSpectrumWarning``. All negative eigenvalues (even smaller ones)
+      are set to zero in all cases.
+
+    - that the eigenvalues are well conditioned. That means, that the
+      eigenvalues are all greater than the maximum eigenvalue divided by 1e12.
+      If this check fails and ``warn_on_zeros=True``, it raises a
+      ``PositiveSpectrumWarning``. All the eigenvalues that are too small are
+      then set to zero.
+
+    Parameters
+    ----------
+    lambdas : array-like, shape (n_eigenvalues,)
+        Array of eigenvalues to check / fix.
+
+    warn_on_zeros : bool, optional (default=False)
+        When this is set to ``True``, a ``PositiveSpectrumWarning`` will be
+        raised when there are extremely small eigenvalues. Otherwise no warning
+        will be raised. Note that in both cases, extremely small eigenvalues
+        will be set to zero.
+
+    Returns
+    -------
+    lambdas_fixed : ndarray, shape (n_eigenvalues,)
+        A fixed validated copy of the array of eigenvalues.
+
+    Examples
+    --------
+    >>> _check_psd_eigenvalues([1, 2])      # nominal case
+    ... # doctest: +NORMALIZE_WHITESPACE
+    array([1, 2])
+    >>> _check_psd_eigenvalues([5, 5j])     # significant imag part
+    ... # doctest: +ELLIPSIS +NORMALIZE_WHITESPACE
+    Traceback (most recent call last):
+        ...
+    ValueError: There are significant imaginary parts in eigenvalues (1.000000
+        of the maximum real part). The matrix is maybe not PSD, or something
+        went wrong with the eigenvalues decomposition.
+    >>> _check_psd_eigenvalues([5, 5e-5j])  # insignificant imag part
+    ... # doctest: +NORMALIZE_WHITESPACE
+    array([5., 0.])
+    >>> _check_psd_eigenvalues([-5, -1])    # all negative
+    ... # doctest: +ELLIPSIS +NORMALIZE_WHITESPACE
+    Traceback (most recent call last):
+        ...
+    ValueError: All eigenvalues are negative (maximum is -1.000000). The matrix
+        is maybe not PSD, or something went wrong with the eigenvalues
+        decomposition.
+    >>> _check_psd_eigenvalues([5, -1])     # significant negative
+    ... # doctest: +NORMALIZE_WHITESPACE
+    array([5, 0])
+    >>> _check_psd_eigenvalues([5, -5e-5])  # insignificant negative
+    ... # doctest: +NORMALIZE_WHITESPACE
+    array([5., 0.])
+    >>> _check_psd_eigenvalues([5, 4e-12])  # bad conditioning (too small)
+    ... # doctest: +NORMALIZE_WHITESPACE
+    array([5., 0.])
+
+    """
+
+    lambdas = np.array(lambdas)
```

Comment:
I feel like if this check is here, it should be `check_array`

Reply:
what check?

Reply:
to me, this is checking that the `lambdas` are convertible to a numpy array, and if they are, but are not an array, it converts them to an array. This logic is usually what we do in `check_array`

Reply:
We also want to copy the input as well. No strong opinion but `check_array` seems overkill to me since in effect what it will do is precisely call `np.array`

Reply:
I think we should avoid using `np.array` in the codebase unless really necessary, so that we can later on handle non-numpy input easier in the future.

Reply:
For user input I agree, but in this specific case less so.

In any case, I tried using check_array but it raises an error on complex values.

Reply:
If it's not data given by the user, then why do we need to convert to an array in the first place?

Reply:
Mostly to make a copy.

It also makes tests easy by providing lists.

Reply:
> Mostly to make a copy.

I'd use `lambdas.copy()` if that's what we need to do.

> It also makes tests easy by providing lists.

Ok. I'm not happy about it, but I'll accept it the way it is then.